### PR TITLE
Slim modules, authority as state params

### PR DESCRIPTION
**Pull request by agouin**
Move authority from a constructor parameter to a param subspace parameter for both `upgrade` and `params` modules.

Defers all possible logic to the SDK params and upgrades modules, only adding the msg server for the new authoritative messages `MsgUpdateParams`, `MsgSoftwareUpgrade`, and `MsgCancelUpgrade`

Cleans up proto files and uses `gogoproto.casttype` to import from the SDK modules for everything except authority.

### CODE REVIEW

File: x/params/keeper/authority.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+func (k Keeper) SetParams(ctx sdk.Context, params proposal.Params) {
+	k.paramSpace.SetParamSet(ctx, &params)
+}
+
+// GetAuthority returns the module authority
+func (k Keeper) GetAuthority(
+	ctx sdk.Context,
+) (authority string) {
+	var res string
+	k.paramSpace.Get(ctx, []byte(types.AuthorityKey), &res)
```
Comment (tac0turtle) [inline x/params/keeper/authority.go:19]: please use storey here. this subspace is mainlkly for params in json encoding

File: proto/cosmos/params/v1beta1/tx.proto [hunk 2]
```diff
@@ -14,11 +15,11 @@ service Msg {
...
 
-  ParameterChangeProposal change_proposal = 2;
+  cosmos.params.v1beta1.ParameterChangeProposal change_proposal = 2 [(gogoproto.casttype) = "github.com/cosmos/cosmos-sdk/x/params/types/proposal.ParameterChangeProposal"];
 }
 
 // UpdateParamsResponse is response type for the Query/Params RPC method.
 message MsgUpdateParamsResponse {
   // param defines the queried parameter.
-  ParamChange param = 1;
+  cosmos.params.v1beta1.ParamChange param = 1 [(gogoproto.casttype) = "github.com/cosmos/cosmos-sdk/x/params/types/proposal.ParamChange"];
```
[the remark this answers is not in the document]
Comment (agouin) [reply proto/cosmos/params/v1beta1/tx.proto:24]: Without doing this at some point to bridge between this override module and the SDK modules, we would have to generate the protos in this repo for all of the SDK protos, which leads to needing all of the business logic for those generated types also.

Do you know of a better way to make sure that the generated protos import from the sdk deps?

File: x/params/keeper/authority.go (hunk 1 above)
Comment (agouin) [reply x/params/keeper/authority.go:19]: Without throwing it in the param space, I couldn't find a great way to use the `MsgUpdateParams` to update the `authority` without introducing another msg type. Any ideas here?

File: proto/cosmos/params/v1beta1/tx.proto (hunk 2 above)
[the remark this answers is not in the document]
Comment (tac0turtle) [reply proto/cosmos/params/v1beta1/tx.proto:24]: ah okay, I guess this is fine. im generally not a fan of using this but I guess it only for a short time till 0.47 is out

File: x/params/keeper/authority.go (hunk 1 above)
Comment (tac0turtle) [reply x/params/keeper/authority.go:19]: how come the normal store doesn't work?

Comment (agouin) [reply x/params/keeper/authority.go:19]: e2e test [here](https://github.com/strangelove-ventures/noble/blob/main/ibctest/paramauthority_test.go)

https://github.com/strangelove-ventures/noble/blob/main/ibctest/paramauthority_test.go#L44-L46
AFAICS an update params proposal and, by extension, this `MsgUpdateParams` following the same implementation, it's only possible to update params within a param subspace, not directly in a store.

Comment (tac0turtle) [reply x/params/keeper/authority.go:19]: its either or, but forgot this was a param, then its chill

**Review by boojamya — APPROVED** [review]

**Pull request merged**